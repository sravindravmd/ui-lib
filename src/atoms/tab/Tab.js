import React, { useState, useEffect, memo } from 'react'
import PropTypes from 'prop-types'
import {
  StyledTabContainer,
  TabHeading,
  TabNavigation,
  StyledTabItem,
} from './Tab.styles'

export const Tab = memo(
  ({ variant, heading, children, navStyle, navClassName, ...props }) => {
    const [tabs, setTabs] = useState([])
    const [activeIndex, setActiveIndex] = useState(0)
    const elements = React.Children.toArray(children).filter(
      ele => typeof ele.type === 'function'
    )

    const Content = React.cloneElement(elements[activeIndex], { title: '' })

    const handleTabChange = index => {
      setActiveIndex(index)
    }

    useEffect(() => {
      const tabs = elements.map((ele, i) => ({
        id: i,
        title: ele.props.title,
        disabled: ele.props.disabled,
      }))
      setTabs(tabs)
    }, [])

    return (
      <StyledTabContainer variant={variant} {...props}>
        {heading && variant === 'horizontal' && (
          <TabHeading>{heading}</TabHeading>
        )}
        <TabNavigation variant={variant}>
          {tabs.length &&
            tabs.map((tab, i) => {
              return (
                <StyledTabItem
                  key={tab.title}
                  onClick={() => handleTabChange(i)}
                  active={activeIndex === i ? true : false}
                  variant={variant}
                  disabled={tab.disabled}
                  style={navStyle}
                  className={navClassName}
                >
                  {variant === 'vertical' ? (
                    <span>{tab.title}</span>
                  ) : (
                    tab.title
                  )}
                </StyledTabItem>
              )
            })}
        </TabNavigation>
        {Content}
      </StyledTabContainer>
    )
  }
)

export const TabItem = ({ children, ...props }) => {
  return <div {...props}>{children}</div>
}

TabItem.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
}

Tab.propTypes = {
  variant: PropTypes.oneOf(['horizontal', 'vertical']),
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
  ]),
  heading: PropTypes.string,
  disabled: PropTypes.bool,
  navStyle: PropTypes.object,
  navClassName: PropTypes.string,
}

Tab.defaultProps = {
  variant: 'horizontal',
  disabled: false,
}

export default Tab
