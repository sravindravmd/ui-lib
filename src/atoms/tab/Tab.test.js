import React from 'react'
import { render, fireEvent, cleanup } from '@testing-library/react'
import { Tab, TabItem } from './Tab'

afterEach(cleanup)

describe('<Tab/>', () => {
  it('renders horizontal tab correctly', () => {
    const { container } = render(
      <Tab heading='ALL Items' variant='horizontal'>
        <TabItem title='Performance'>performance</TabItem>
        <TabItem title='Recent Activity'>Recent Activity</TabItem>
      </Tab>
    )
    expect(container).toMatchSnapshot()
  })

  it('renders vertical tab correctly', () => {
    const { container } = render(
      <Tab variant='vertical'>
        <TabItem title='Performance'>performance</TabItem>
        <TabItem title='Recent Activity'>Recent Activity</TabItem>
      </Tab>
    )
    expect(container).toMatchSnapshot()
  })

  it('renders horizontal tab with TabItem disabled', () => {
    const { container } = render(
      <Tab heading='Tooltip' variant='horizontal'>
        <TabItem title='Performance' disabled>
          performance
        </TabItem>
        <TabItem title='Recent Activity'>Recent Activity</TabItem>
      </Tab>
    )
    expect(container).toMatchSnapshot()
  })

  it('render selected tab content correctly', () => {
    const { getByText } = render(
      <Tab heading='Tooltip' variant='horizontal'>
        <TabItem title='Performance'>Content for performance</TabItem>
        <TabItem title='Recent Activity'>Content for Recent activity</TabItem>
      </Tab>
    )
    const tabItem = getByText('Performance')
    fireEvent.click(tabItem)
    expect(getByText('Content for performance')).toBeInTheDocument()
  })
})
