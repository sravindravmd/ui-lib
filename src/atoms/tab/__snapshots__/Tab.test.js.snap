// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab/> renders horizontal tab correctly 1`] = `
<div>
  <div
    class="sc-bdfBwQ dmxWwQ"
  >
    <div
      class="sc-gsTCUz iFpzxu"
    >
      ALL Items
    </div>
    <div
      class="sc-dlfnbm btJzNe"
    >
      <div
        class="sc-hKgILt bWZNvk"
      >
        Performance
      </div>
      <div
        class="sc-hKgILt kwxITU"
      >
        Recent Activity
      </div>
    </div>
    <div
      title=""
    >
      performance
    </div>
  </div>
</div>
`;

exports[`<Tab/> renders horizontal tab with TabItem disabled 1`] = `
<div>
  <div
    class="sc-bdfBwQ dmxWwQ"
  >
    <div
      class="sc-gsTCUz iFpzxu"
    >
      Tooltip
    </div>
    <div
      class="sc-dlfnbm btJzNe"
    >
      <div
        class="sc-hKgILt iiZBkb"
        disabled=""
      >
        Performance
      </div>
      <div
        class="sc-hKgILt kwxITU"
      >
        Recent Activity
      </div>
    </div>
    <div
      disabled=""
      title=""
    >
      performance
    </div>
  </div>
</div>
`;

exports[`<Tab/> renders vertical tab correctly 1`] = `
<div>
  <div
    class="sc-bdfBwQ kvrjgw"
  >
    <div
      class="sc-dlfnbm frgbOH"
    >
      <div
        class="sc-hKgILt iTCrqA"
      >
        <span>
          Performance
        </span>
      </div>
      <div
        class="sc-hKgILt UdVBa"
      >
        <span>
          Recent Activity
        </span>
      </div>
    </div>
    <div
      title=""
    >
      performance
    </div>
  </div>
</div>
`;
